### PR TITLE
Handling stdout on Windows

### PR DESCRIPTION
**Pull request by peterbud**
This PR is aiming to solve the small problem we had with `stdout` on Windows, namely that even simple command line arguments like `--help` or `--version` did not show anything on the command line console /command prompt as all `stdout` was already redirected to a log file, hence the user did not see the result.

This change is adding back the console subsystem to the compiler and linker flags, (while keeping the `-mwindows`) and adds a simple logic at startup, that for simple command line arguments, like `-h`, `--help` and `--version` the `stdout` won't be redirected, so the user would see the output, while for other cases the existing redirection to the log file is still working.

Also I have extended the `--help`, with showing the location of the debug log file to help the users to find this if its needed.

This way both file logging and console output is working.

### CODE REVIEW

File: src/common/darktable.c [hunk 1]
```diff
@@ -112,6 +117,9 @@ static int usage(const char *argv0)
   printf("  --configdir <user config directory>\n");
   printf("  -d {all,cache,camctl,camsupport,control,dev,fswatch, input,lighttable,\n");
   printf("      lua, masks,memory,nan,opencl, perf,pwstorage,print,sql}\n");
+#ifdef _WIN32
+  printf("     Note: debug log will be written to this file: %s\n", logfile);
```
Comment (houz) [inline src/common/darktable.c]: That shouldn't be part of the `-d` option as it affects all output, not just the debug log.

File: src/common/darktable.c [hunk 2]
```diff
@@ -103,6 +103,11 @@ darktable_t darktable;
 
 static int usage(const char *argv0)
 {
+#ifdef _WIN32
+  char *logdir = g_build_filename(g_get_user_cache_dir(), "darktable", NULL);
```
Comment (houz) [inline src/common/darktable.c]: What is `logdir` created for?

File: src/CMakeLists.txt [hunk 3]
```diff
@@ -817,7 +817,7 @@ endif(APPLE)
 target_link_libraries(darktable lib_darktable)
 
 if (WIN32)
-  set_target_properties(darktable PROPERTIES LINK_FLAGS "-mwindows -Wl,-subsystem,windows")
+  set_target_properties(darktable PROPERTIES LINK_FLAGS "-mconsole -mwindows -Wl,-subsystem,console")
```
Comment (houz) [inline src/CMakeLists.txt]: And this allows printing to the `cmd.exe` console while not showing a console when starting the program?

Comment (peterbud) [reply src/CMakeLists.txt]: Yes :) You can give a try

File: src/common/darktable.c (hunk 2 above)
Comment (peterbud) [reply src/common/darktable.c]: To be able to print out later where the user can find the logfile, as its pretty hidden.

File: src/common/darktable.c (hunk 1 above)
Comment (peterbud) [reply src/common/darktable.c]: I can move it to the end of the usage text, no problem.

File: src/common/darktable.c (hunk 2 above)
Comment (houz) [reply src/common/darktable.c]: I must be blind, I don't see it being used anywhere.

Comment (peterbud) [reply src/common/darktable.c]: Next line is using that `logdir`
`char *logfile = g_build_filename(*logdir*, "darktable-log.txt", NULL);`
Actually I have copied that from [here](https://github.com/darktable-org/darktable/blob/master/src/main.c#L43)
You mean maybe to use only one `g_build_filename()`?

Comment (houz) [reply src/common/darktable.c]: Yes. :-)

File: src/CMakeLists.txt (hunk 3 above)
Comment (houz) [reply src/CMakeLists.txt]: I just tried on Windows 7. When double clicking darktable.exe in the explorer I get the dreaded terminal window.

Comment (peterbud) [reply src/CMakeLists.txt]: Indeed. There was no terminal windows when I was launching from command prompt. I have a solution, I'll push it shortly.

File: src/common/darktable.c (hunk 2 above)
Comment (peterbud) [reply src/common/darktable.c]: This is also done with the latest commit

File: src/CMakeLists.txt (hunk 3 above)
Comment (peterbud) [reply src/CMakeLists.txt]: So I have added the `FreeConsole()` call to the part when we decide to use log to file, and hence no more console is needed. This results that at startup you see a console window flash for a second, then disappears, and darktable's main window will appear, and only that window will be visible.
I have tested the following scenarios:
* Launching w/o arguments from command prompt: works, only dt's main window will be visible
* Launching with `--help` from command prompt: works, you see in the console the usage text
* Launching w/o arguments from explorer by double click: works, only dt's main window will be visible
* I have even tested launching with `--help` from explorer by creating a shortcut and double click on that: that is rather useless, but it displays only a console window, with a usage text, but then it closes, so not too much point IMHO.

**Comment by houz** [comment]
The flashing window isn't ideal, but it's much better than what we had before. Thank you for looking into this.

**Comment by peterbud** [comment]
I'll look later what would it take in the opposite direction, ie no console by default, and then for `--help`, `--version` add a console by hand.

**Pull request merged**